### PR TITLE
Add MPU6050 library and sample code

### PR DESCRIPTION
**Pull request by xeonqq**

### CODE REVIEW

**Review by mikee47 — COMMENTED** [review]
A few thoughts

**Comment by xeonqq** [comment]
After a closer examination, I realized the quaternion related code is not used in this part of the library (see [here](https://github.com/SmingHub/Sming/pull/2655/files#diff-ecc89b792b3b1fdbd6efe09c73dc777fc73ea29968ae8366a2c1ca37d1f3dc6fR6)). So I simply deleted it to keep the library lean.

The other comments of yours I have addressed. Thank you for the feedback. Please have a look again~

**Comment by slaff** [comment]
@mikee47 The build with IDF v5.0 is failing. I was able to reproduce the issue on my machine and it seems to not be related to this PR. Can you check if we have to downgrade/upgrade the pyparse version?
```
During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen.py", line 183, in <module>
    main()
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen.py", line 159, in main
    mapping_rules = generation_model.generate(sections_infos)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 514, in generate
    root_node.insert(entity, sections, target, flags, entities)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 241, in insert
    self.child_placement(entity, sections, target, flags, sections_db)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 233, in child_placement
    child.insert(entity, sections, target, flags, sections_db)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 241, in insert
    self.child_placement(entity, sections, target, flags, sections_db)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 233, in child_placement
    child.insert(entity, sections, target, flags, sections_db)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 241, in insert
    self.child_placement(entity, sections, target, flags, sections_db)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/generation.py", line 296, in child_placement
    found_sections = sections_db.get_sections(self.parent.name, self.name)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/entity.py", line 197, in get_sections
    obj = self._match_obj(archive, obj)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/entity.py", line 183, in _match_obj
    objs = self.get_objects(archive)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/entity.py", line 176, in get_objects
    self._process_archive(archive)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/entity.py", line 164, in _process_archive
    parsed = self._get_infos_from_file(stored)
  File "/opt/esp-idf-5.0/tools/ldgen/ldgen/entity.py", line 154, in _get_infos_from_file
    raise ParseException('Unable to parse section info file ' + info.filename + '. ' + p.msg)
pyparsing.exceptions.ParseException: Unable to parse section info file /home/runner/work/Sming/Sming/Sming/out/Esp32/esp32/debug/build/esp32/sdk/esp-idf/esp_event/libesp_event.a. Expected end of text  (at char 0), (line:1, col:1)
ninja: build stopped: subcommand failed.
Error: Process completed with exit code 2.
```

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 1]
```diff
@@ -0,0 +1,3650 @@
...
+{
+}
+
+/** Power on and prepare for general usage.
+ * This will activate the device and take it out of sleep mode (which must be
+ * done after start-up). This function also sets both the accelerometer and the
+ * gyroscope to their most sensitive settings, namely +/- 2g and +/- 250
+ * degrees/sec, and sets the clock source to use the X Gyro for reference, which
+ * is slightly better than the default internal clock source.
+ */
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.cpp:66]: If you want to make use of  comments to appear in the documention they should be in the header file.

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 2]
```diff
@@ -0,0 +1,837 @@
...
+
+#undef pgm_read_byte
+#define pgm_read_byte(addr) (*static_cast<const unsigned char*>(addr))
+
+//#define PROGMEM /* empty */
+//#define pgm_read_byte(x) (*(x))
+//#define pgm_read_word(x) (*(x))
+//#define pgm_read_float(x) (*(x))
+//#define PSTR(STR) STR
+
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:55]: Remove. That redefinition of `pgm_read_byte` will cause issues for esp8266.

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 3]
```diff
@@ -0,0 +1,837 @@
...
+AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+THE SOFTWARE.
+===============================================
+*/
+
+#ifndef _MPU6050_H_
+#define _MPU6050_H_
+
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:40]: `#pragma once` preferred

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 4]
```diff
@@ -0,0 +1,837 @@
...
+	uint8_t getDMPConfig2();
+	void setDMPConfig2(uint8_t config);
+
+	void CalibrateGyro(uint8_t Loops = 15);  // Fine tune after setting offsets with less Loops.
+	void CalibrateAccel(uint8_t Loops = 15); // Fine tune after setting offsets with less Loops.
+	void PID(uint8_t ReadAddress, float kP, float kI, uint8_t Loops); // Does the
+
+private:
+	uint8_t devAddr;
+	uint8_t buffer[14] = {0};
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:834]: Remove. Methods should use local variables.

File: Sming/Libraries/MPU6050/README.md [hunk 5]
```diff
@@ -0,0 +1,8 @@
+# MPU6050 Six-Axis (Gyro + Accelerometer)
+
+Based on code from [jrowberg/i2cdevlib](https://github.com/jrowberg/i2cdevlib/tree/master/ESP32_ESP-IDF/components/MPU6050). Most of the code is the same, except:
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/README.md:3]: Maybe date and which commit this is based on.

Can also add doxygen documentation, but requires this file to be renamed to `README.rst` and heading/url adjusted accordingly.

```
API Documentation
-----------------

.. doxygenclass:: MPU6050
   :members:

```

File: samples/Accel_Gyro_MPU6050/app/application.cpp [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+}
+
+void mainLoop()
+{
+	int16_t ax, ay, az;
+	int16_t gx, gy, gz;
+	mpu.getMotion6(&ax, &ay, &az, &gx, &gy, &gz);
+	printAccelGyro(ax, ay, az, gx, gy, gz);
+}
+
```
Comment (mikee47) [inline samples/Accel_Gyro_MPU6050/app/application.cpp:20]:
```
void mainLoop()
{
	MPU6050::Motion6 motion;
	mpu.getMotion6(motion);
        Serial << motion << endl;
}
```

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 7]
```diff
@@ -0,0 +1,3650 @@
...
+/** Get the auxiliary I2C supply voltage level.
+ * When set to 1, the auxiliary I2C bus high logic level is VDD. When cleared to
+ * 0, the auxiliary I2C bus high logic level is VLOGIC. This does not apply to
+ * the MPU-6000, which does not have a VLOGIC pin.
+ * @return I2C supply voltage level (0=VLOGIC, 1=VDD)
+ */
+uint8_t MPU6050::getAuxVDDIOLevel()
+{
+	I2Cdev::readBit(devAddr, MPU6050_RA_YG_OFFS_TC, MPU6050_TC_PWR_MODE_BIT, buffer);
+	return buffer[0];
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.cpp:95]: Use stack for buffer.

```
uint8_t b;
I2Cdev::readBit(devAddr, MPU6050_RA_YG_OFFS_TC, MPU6050_TC_PWR_MODE_BIT, &b);
return b;
```

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 8]
```diff
@@ -0,0 +1,3650 @@
...
+ * For a diagram of the gyroscope and accelerometer signal paths, see Section 8
+ * of the MPU-6000/MPU-6050 Product Specification document.
+ *
+ * @return Current sample rate
+ * @see MPU6050_RA_SMPLRT_DIV
+ */
+uint8_t MPU6050::getRate()
+{
+	I2Cdev::readByte(devAddr, MPU6050_RA_SMPLRT_DIV, buffer);
+	return buffer[0];
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.cpp:134]: Use stack for buffer.

```
uint8_t b;
I2Cdev::readByte(devAddr, MPU6050_RA_SMPLRT_DIV, &b);
return b;
```

or use private helper:

```
return readByte(MPU6050_RA_SMPLRT_DIV);
```

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 9]
```diff
@@ -0,0 +1,3650 @@
...
+ * 6            | ACCEL_YOUT_L[0]
+ * 7            | ACCEL_ZOUT_L[0]
+ * </pre>
+ *
+ * @return FSYNC configuration value
+ */
+uint8_t MPU6050::getExternalFrameSync()
+{
+	I2Cdev::readBits(devAddr, MPU6050_RA_CONFIG, MPU6050_CFG_EXT_SYNC_SET_BIT, MPU6050_CFG_EXT_SYNC_SET_LENGTH, buffer);
+	return buffer[0];
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.cpp:178]: `return readBits(MPU6050_RA_CONFIG, MPU6050_CFG_EXT_SYNC_SET_BIT, MPU6050_CFG_EXT_SYNC_SET_LENGTH);`

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 10]
```diff
@@ -0,0 +1,3650 @@
...
+
+/** Get self-test factory trim value for accelerometer Y axis.
+ * @return factory trim value
+ * @see MPU6050_RA_SELF_TEST_Y
+ */
+uint8_t MPU6050::getAccelYSelfTestFactoryTrim()
+{
+	I2Cdev::readByte(devAddr, MPU6050_RA_SELF_TEST_Y, &buffer[0]);
+	I2Cdev::readByte(devAddr, MPU6050_RA_SELF_TEST_A, &buffer[1]);
+	return (buffer[0] >> 3) | ((buffer[1] >> 2) & 0x03);
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.cpp:295]: Using helpers:

```
uint8_t y = readByte(MPU6050_RA_SELF_TEST_Y);
uint8_t a = readByte(MPU6050_RA_SELF_TEST_A);
return (y >> 3) | ((a >> 2) & 0x03);
```

File: Sming/Libraries/MPU6050/component.mk [hunk 11]
```diff
@@ -0,0 +1 @@
+COMPONENT_DEPENDS := I2Cdev
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/component.mk:1]: To make doxygen comments available (from .h files) add `COMPONENT_DOXYGEN_INPUT := .`

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 12]
```diff
@@ -0,0 +1,837 @@
...
+#define MPU6050_BANKSEL_MEM_SEL_LENGTH 5
+
+#define MPU6050_WHO_AM_I_BIT 6
+#define MPU6050_WHO_AM_I_LENGTH 6
+
+#define MPU6050_DMP_MEMORY_BANKS 8
+#define MPU6050_DMP_MEMORY_BANK_SIZE 256
+#define MPU6050_DMP_MEMORY_CHUNK_SIZE 16
+
+// note: DMP code memory blocks defined at end of header file
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:434]: Remove (or edit to make sense)

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 13]
```diff
@@ -0,0 +1,837 @@
...
+	void CalibrateGyro(uint8_t Loops = 15);  // Fine tune after setting offsets with less Loops.
+	void CalibrateAccel(uint8_t Loops = 15); // Fine tune after setting offsets with less Loops.
+	void PID(uint8_t ReadAddress, float kP, float kI, uint8_t Loops); // Does the
+
+private:
+	uint8_t devAddr;
+	uint8_t buffer[14] = {0};
+};
+
+#endif /* _MPU6050_H_ */
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:837]: Remove (#pragma once)

**Review by mikee47 — COMMENTED** [review]
This is an example of code which got written with a 'C' viewpoint rather than C++. More robust, easier to manage and safer to use references instead of pointers, structures instead of long parameter lists, etc. Hopefully this all makes sense?

**Comment by xeonqq** [comment]
Thanks for the review, I agree the old code which I copy/paste is not to the standard of Sming. I will do the refactoring when I find time.  Currently I want to first prototype the balance robot I am working on:)

**Comment by slaff** [comment]
@xeonqq are you ready with all changes to this PR?

**Comment by xeonqq** [comment]
> @xeonqq are you ready with all changes to this PR?

No, not yet, there are still many comments I didn't address yet. Quite some workload. I will remove the wip tag in the PR title once I am ready for another review.

**Comment by xeonqq** [comment]
@mikee47 ready for review again. Sorry for the delayed PR.
I in the end deleted Calibration and Memory Block related code for simplicity and code quality reason, anyway that is not frequently used functionality.
I fail one of the CI checks, it's about some clock cycles I am not sure if it's related to my change.

**Comment by mikee47** [comment]
> I fail one of the CI checks, it's about some clock cycles I am not sure if it's related to my change.

Does that sometimes, host emulator isn't perfect :-( Re-running the test

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 14]
```diff
@@ -0,0 +1,3444 @@
...
+#define MPU6050_BANKSEL_MEM_SEL_BIT 4
+#define MPU6050_BANKSEL_MEM_SEL_LENGTH 5
+
+#define MPU6050_WHO_AM_I_BIT 6
+#define MPU6050_WHO_AM_I_LENGTH 6
+
+#define MPU6050_DMP_MEMORY_BANKS 8
+#define MPU6050_DMP_MEMORY_BANK_SIZE 256
+#define MPU6050_DMP_MEMORY_CHUNK_SIZE 16
+
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:415]: Most (if not all) of these macros are for internal use by the class. Suggest moving them into a separate header file so they don't clutter the global namespace.

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 15]
```diff
@@ -0,0 +1,3444 @@
...
+	{
+	}
+
+	/** Specific address constructor.
+         * @param address I2C address
+         * @see MPU6050_DEFAULT_ADDRESS
+         * @see MPU6050_ADDRESS_AD0_LOW
+         * @see MPU6050_ADDRESS_AD0_HIGH
+         */
+	MPU6050(uint8_t address) : devAddr{address}
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:444]: `uint8_t` -> `SlaveId`

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 16]
```diff
@@ -0,0 +1,3444 @@
...
+#define MPU6050_WHO_AM_I_LENGTH 6
+
+#define MPU6050_DMP_MEMORY_BANKS 8
+#define MPU6050_DMP_MEMORY_BANK_SIZE 256
+#define MPU6050_DMP_MEMORY_CHUNK_SIZE 16
+
+class MPU6050
+{
+public:
+	struct Motion3 {
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/MPU6050.h:419]: using SlaveId = uint8_t;

File: Sming/Libraries/MPU6050/README.rst [hunk 17]
```diff
@@ -0,0 +1,14 @@
+API Documentation
+-----------------
+
+.. doxygenclass:: MPU6050
+   :members:
+
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/README.rst:6]: Move this to the bottom of the file for consistency - it gets expanded into a lot of information.

File: Sming/Libraries/MPU6050/README.rst [hunk 18]
```diff
@@ -0,0 +1,14 @@
...
+# MPU6050 Six-Axis (Gyro + Accelerometer)
+Based on code from [jrowberg/i2cdevlib](https://github.com/jrowberg/i2cdevlib/tree/master/ESP32_ESP-IDF/components/MPU6050) @ 605a740. Most of the code is the same, except:
+
+-   Removed MPU6050::ReadRegister function due to incompatibility. It is also not used anywhere in the original code.
+-   MPU6050_6Axis_MotionApps20.h and MPU6050_9Axis_MotionApps41.h are not included due to deps to freeRTOS. helper_3dmath.h is also not included since it is only used in the above mentioned files.
+-   Removed map function in favor of the Sming built-in one.
+-   Adapted include path, coding style and applied clangformat
+-   Deleted Calibration and Memory Block related code for code quality reason
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/README.rst:14]: Have a look at other libraries/Components for general layout. Heading must be double-underlined (level 1) and give a good name for the library. For example:

```
MPU6050 Gyro / Accelerometer
============================
```

Follow this with the description. Note that `.rst` files (restructured text) syntax differs from markdown so remember to fix links, e.g.

```
`jrowberg/i2cdevlib<https://github.com/jrowberg/i2cdevlib/tree/master/ESP32_ESP-IDF/components/MPU6050>`
```

And put the API section last.

**Review by mikee47 — CHANGES_REQUESTED** [review]
You've inadvertently reverted the `FlashString` and `IFS` libraries! Other than that, and the documentation change, looks great! I assume it's all working and tested on actual hardware?

**Comment by xeonqq** [comment]
> You've inadvertently reverted the `FlashString` and `IFS` libraries! Other than that, and the documentation change, looks great! I assume it's all working and tested on actual hardware?

I just realized that too, and reverted the unintended change of the submodules. 

I tested on esp8266 and mpu6050, but only the functions used in the sample program. So I can't say for sure it is bullet proof.
And tested assertion is working as expected.

File: Sming/Libraries/MPU6050/README.rst [hunk 19]
```diff
@@ -0,0 +1,20 @@
+MPU6050 Gyro / Accelerometer
+=============================
+
+   :members:
+
+# MPU6050 Six-Axis (Gyro + Accelerometer)
+Based on code from `jrowberg/i2cdevlib <https://github.com/jrowberg/i2cdevlib/tree/master/ESP32_ESP-IDF/components/MPU6050>` @ 605a740. Most of the code is the same, except:
+
```
Comment (mikee47) [inline Sming/Libraries/MPU6050/README.rst:8]: Almost... remove `:members:` and line starting '#'. Also the URL requires __ at the end!

```
Based on code from `jrowberg/i2cdevlib <https://github.com/jrowberg/i2cdevlib/tree/master/ESP32_ESP-IDF/components/MPU6050>`__ @ 605a740. Most of the code is the same, except:
```

File: Sming/Libraries/MPU6050/README.rst (hunk 18 above)
Comment (mikee47) [reply Sming/Libraries/MPU6050/README.rst]: Yeah, takes a bit of getting used to but has advantages to markdown! See https://sming.readthedocs.io/en/latest/information/develop/documentation.html

File: Sming/Libraries/MPU6050/MPU6050.cpp [hunk 20]
```diff
@@ -0,0 +1,323 @@
+// I2Cdev library collection - MPU6050 I2C device class
+// Based on InvenSense MPU-6050 register map document rev. 2.0, 5/19/2011
+// (RM-MPU-6000A-00) 8/24/2011 by Jeff Rowberg <jeff@rowberg.net> Updates should
+// (hopefully) always be available at https://github.com/jrowberg/i2cdevlib
+//
+// Changelog:
+//     ... - ongoing debug release
+
+// NOTE: THIS IS ONLY A PARIAL RELEASE. THIS DEVICE CLASS IS CURRENTLY
```
Comment (slaff) [inline Sming/Libraries/MPU6050/MPU6050.cpp:9]: Fix typo, replace  `PARIAL` with `PARTIAL`.

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 21]
```diff
@@ -0,0 +1,3446 @@
+// Based on InvenSense MPU-6050 register map document rev. 2.0, 5/19/2011
+// (RM-MPU-6000A-00)
+// Based On https://github.com/jrowberg/i2cdevlib
+
+// NOTE: THIS IS ONLY A PARIAL RELEASE. THIS DEVICE CLASS IS CURRENTLY
```
Comment (slaff) [inline Sming/Libraries/MPU6050/MPU6050.h:5]: Fix typo, replace  `PARIAL` with `PARTIAL`.

File: Sming/Libraries/MPU6050/MPU6050.h [hunk 22]
```diff
@@ -0,0 +1,3446 @@
...
+	uint8_t devAddr;
+};
+
+template <typename T> T MPU6050::readReg(uint8_t regAddr)
+{
+	static_assert(std::is_fundamental<T>::value, "T must be an fundamental type.");
+
+	const auto sz = sizeof(T);
+	uint8_t buffer[sz] = {0};
+	//data follow big endien convention
```
Comment (slaff) [inline Sming/Libraries/MPU6050/MPU6050.h:3430]: Fix typo, replace `endien` with `endian`.

**Review by slaff — CHANGES_REQUESTED** [review]
Fix small typos.

**Review by mikee47 — APPROVED** [review]

File: Sming/Libraries/MPU6050/MPU6050.h (hunk 22 above)
Comment (xeonqq) [reply Sming/Libraries/MPU6050/MPU6050.h]: done

**Review by slaff — APPROVED** [review]

**Pull request merged**